assets/js/ubi.js

```diff
@@ -0,0 +1,300 @@
+function guiid() {
+  let id = '123456-insecure';
+  try {
+    id = crypto.randomUUID();
+  }
+  catch(error){
+    console.warn('tried to generate a guiid in an insecure context');
+    id ='10000000-1000-4000-8000-100000000000'.replace(/[018]/g, c =>
+    (c ^ crypto.getRandomValues(new Uint8Array(1))[0] & 15 >> c / 4).toString(16));
+  }
+  return id;
+};
+
+export function hash(str, seed=42) {
+  let h1 = 0xdeadbeef ^ seed, h2 = 0x41c6ce57 ^ seed;
+  for(let i = 0, ch; i < str.length; i++) {
+    ch = str.charCodeAt(i);
+    h1 = Math.imul(h1 ^ ch, 2654435761);
+    h2 = Math.imul(h2 ^ ch, 1597334677);
+  }
+  h1  = Math.imul(h1 ^ (h1 >>> 16), 2246822507);
+  h1 ^= Math.imul(h2 ^ (h2 >>> 13), 3266489909);
+  h2  = Math.imul(h2 ^ (h2 >>> 16), 2246822507);
+  h2 ^= Math.imul(h1 ^ (h1 >>> 13), 3266489909);
+
+  return 4294967296 * (2097151 & h2) + (h1 >>> 0);
+};
+
+
+/**
+ * In place of true authentication, this makes a hash out of the user's IP address and browser
```

Comment:
Employees of large organizations usually use versions of the browsers imposed on to them and all of their connections would be from the IP of the organization. Does this mean they would all be recognized as the same user?

Reply:
Yep.  It was the best we could think of without having any access to the webserver code.  Part of User Behavior Insights is tracking individual users (and their behaviors).  This method is obviously not guaranteed to be unique, but it's unique-ish to the ip, and should provide some individual signal.  We figured using the google api was reasonable since you're already using google analytics.
However, since you are far more familiar with the production environment, if you can think of other ways to track individual users, we'll be happy to include that instead.